crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,408 @@
+                    self.reset_with_on_chain_state(jwks)
+                },
+                (_sender, msg) = rpc_req_rx.select_next_some() => {
+                    self.process_peer_request(msg)
+                },
+                qc_update = qc_update_rx.select_next_some() => {
+                    self.process_quorum_certified_update(qc_update)
+                },
+                (issuer, jwks) = local_observation_rx.select_next_some() => {
+                    let jwks = jwks.into_iter().map(JWKMoveStruct::from).collect();
+                    self.process_new_observation(issuer, jwks)
+                },
+                ack_tx = close_rx.select_next_some() => {
+                    self.tear_down(ack_tx.ok()).await
+                }
+            };
+
+            if let Err(e) = handle_result {
+                error!("[JWK] handling_err={}", e);
+            }
+        }
+    }
+
+    async fn tear_down(&mut self, ack_tx: Option<oneshot::Sender<()>>) -> Result<()> {
+        self.stopped = true;
+        let futures = std::mem::take(&mut self.jwk_observers)
+            .into_iter()
+            .map(JWKObserver::shutdown)
+            .collect::<Vec<_>>();
+        join_all(futures).await;
+        if let Some(tx) = ack_tx {
+            let _ = tx.send(());
+        }
+        Ok(())
+    }
+
+    /// Triggered by an observation thread periodically.
+    pub fn process_new_observation(
+        &mut self,
+        issuer: Issuer,
+        jwks: Vec<JWKMoveStruct>,
+    ) -> Result<()> {
+        let state = self.states_by_issuer.entry(issuer.clone()).or_default();
+        state.observed = Some(jwks.clone());
+        if state.observed.as_ref() != state.on_chain.as_ref().map(ProviderJWKs::jwks) {
+            let observed = ProviderJWKs {
+                issuer: issuer.clone(),
+                version: state.on_chain_version() + 1,
+                jwks,
+            };
+            let signature = self
+                .consensus_key
+                .sign(&observed)
+                .map_err(|e| anyhow!("crypto material error occurred duing signing: {}", e))?;
+            let abort_handle = self.certified_update_producer.start_produce(
+                self.epoch_state.clone(),
+                observed.clone(),
+                self.qc_update_tx.clone(),
+            );
+            state.consensus_state = ConsensusState::InProgress {
+                my_proposal: ObservedUpdate {
+                    author: self.my_addr,
+                    observed: observed.clone(),
+                    signature,
+                },
+                abort_handle_wrapper: QuorumCertProcessGuard::new(abort_handle),
+            };
+            info!("[JWK] update observed, update={:?}", observed);
+        }
+
+        Ok(())
+    }
+
+    /// Invoked on start, or on on-chain JWK updated event.
+    pub fn reset_with_on_chain_state(&mut self, on_chain_state: AllProvidersJWKs) -> Result<()> {
```

Comment:
The name is too similar with `reset_with_onchain_state`.. change it?

Reply:
deleted the other one instead. that was not used.